src/shell/shell_cmds.cpp

```diff
@@ -2242,3 +2243,290 @@ void DOS_Shell::CMD_VOL(char* args)
+					continue_after_stop = false;
+				}
+				args++;
+			}
+			break;
+		case ',':
+			stop_next_letter = true;
+			continue_after_stop = true;
+			if (!word.empty()) {
+				sources.emplace_back(std::move(word));
+				word = {};
+			}
+			args++;
+			break;
+		case 0:
+			if (!word.empty()) {
+				sources.emplace_back(std::move(word));
+			}
+			parsing = false;
+			break;
+		default:
+			if (stop_next_letter) {
+				parsing             = continue_after_stop;
+				continue_after_stop = false;
+				stop_next_letter    = false;
+			} else {
+				word.push_back(*args++);
+			}
+		}
+	}
+	return sources;
+}
+
+static std::string cmd_move_parse_destination(const char*& args)
+{
+	std::string destination;
+	bool parsing      = true;
+	bool inside_quote = false;
+
+	while (parsing) {
+		switch (*args) {
+		case '"':
+			inside_quote = !inside_quote;
+			args++;
+			break;
+		case ' ':
+			if (inside_quote) {
+				destination.push_back(*args++);
+			} else {
+				args++;
+				parsing = false;
+			}
+			break;
+		case '\0': parsing = false; break;
+		default: destination.push_back(*args++);
+		}
+	}
+	return destination;
+}
+
+struct cmd_move_arguments {
+	std::vector<std::string> sources = {};
+	std::string destination          = {};
+	const char* error                = nullptr;
+};
+
+cmd_move_arguments cmd_move_parse_arguments(const char* args)
+{
+	cmd_move_arguments ret;
+	ret.sources = cmd_move_parse_sources(args);
+	if (ret.sources.empty()) {
+		ret.error = "SHELL_MISSING_PARAMETER";
+		return ret;
+	}
+	ret.destination = cmd_move_parse_destination(args);
+	if (ret.destination.empty()) {
+		ret.error = "SHELL_MISSING_PARAMETER";
+		return ret;
+	}
+	// Check for extra arguments
+	while (*args) {
+		if (*args++ != ' ') {
+			ret.error = "SHELL_TOO_MANY_PARAMETERS";
+			return ret;
+		}
+	}
+	return ret;
+}
+
+void DOS_Shell::CMD_MOVE(char* args)
+{
+	HELP("MOVE");
+
+	// TODO: Add support for these flags along with the COPYCMD environment
+	// variable (overwrite prompts) Also currently ignored by CMD_COPY
+	(void)ScanCMDBool(args, "Y");
+	(void)ScanCMDBool(args, "-Y");
+
+	char* rem = ScanCMDRemain(args);
+	if (rem) {
+		WriteOut(MSG_Get("SHELL_ILLEGAL_SWITCH"), rem);
+		return;
+	}
+
+	cmd_move_arguments parsed = cmd_move_parse_arguments(args);
+	if (parsed.error) {
+		WriteOut(MSG_Get(parsed.error));
+		return;
+	}
+
+	char temp[DOS_PATHLENGTH];
+	if (!DOS_Canonicalize(parsed.destination.c_str(), temp)) {
+		WriteOut(MSG_Get("SHELL_FILE_NOT_FOUND"),
+		         parsed.destination.c_str());
+		return;
+	}
+	std::string canonical_destination = temp;
+
+	// Command uses dta so set it to our internal dta
+	const RealPt save_dta = dos.dta();
+	dos.dta(dos.tables.tempdta);
+	DOS_DTA dta(dos.dta());
+
+	// Each raw source can be a wildcard (*.*) containing multiple real
+	// sources.
+	std::vector<std::string> final_sources;
+	for (const std::string& source : parsed.sources) {
+		if (!DOS_FindFirst(source.c_str(), 0xffff & ~DOS_ATTR_VOLUME)) {
```

Comment:
Could we do better here with the bit manipulations? Look into `bitops.h` and perhaphs `bit_view.h`.

Reply:
Well, it's marked obsolete here and there's a newer bit_view implementation:  https://github.com/dosbox-staging/dosbox-staging/blob/d64ad72581e08e0c48b9dcc9d5da7703f03fb8b8/include/dos_system.h#L44-L75

But from grepping the callers of this specific function, all except one (in `program_tree)` use this old way.  It also takes in a `uint16_t` even though it's supposed to be a 6 bit field?  I'm not really sure what's going on here.  The function probably needs  a refactor at some point.

Reply:
Yeah, it's a mess 😩 In `DOS_FindFirst` the `uint16_t attr` only gets compared to `DOS_ATTR_VOLUME` (which is `uint8_t`) and then it's cast to `uint8_t` anyway (without `check_cast`) here:

```cpp
dta.SetupSearch(drive,(uint8_t)attr,pattern);
```

So it is safe to use the `FatAttributeFlags` bit_view; please do that and then just cast the result to `uint16_t` on the way in... It's a start 🤷🏻 

Btw, I've tried changing `attr` to `uint8_t` in `DOS_FindFirst`, but it results in too many warnings. I agree that cleaning this up should be a separate piece of work. If you could create a ticket for it to migrate to `FatAttributeFlags`, that would be much appreciated!

<img width="1525" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/2c4014e3-9da2-4a6a-bcb7-02fcd16d268e">

Reply:
How do you suggest using it exactly?  I'm not really familiar with this type of interface.  In this case, it wants every bit set except for volume.  So do you have to do:

``` cpp
FatAttributeFlags attr;
attr.read_only = true;
attr.hidden = true;
attr.system = true;
attr.volume = false;
attr.directory = true;
attr.archive = true;
attr.device = true;
attr.unused = idk... it's unsued?
```
Seems a bit overly verbose to me?

Reply:
Yes, that's too verbose. These two variants both work:

```cpp
FatAttributeFlags attr = {0};
bit::flip_all(attr._data);
attr.volume.flip();
```
```cpp
FatAttributeFlags attr = {0xff};
attr.volume.flip();
```

I'd say use the second form for now, then we can look into improving `bit_view` for such operations later (to init with all bits set without having to specify `0xff`).

Reply:
Thanks.  Yeah, that looks nicer.  Fixed.

Reply:
No problem! Are you going to create a ticket for refactoring everything to move to 'FatAttributeFlags'? If not that's fine, I just need to know then I'll do it myself later 😄

Reply:
There you go:  https://github.com/dosbox-staging/dosbox-staging/issues/2663

Reply:
Cheers @weirddan455 !